saleor/discount/utils.py

```diff
@@ -677,6 +840,107 @@ def _create_or_update_checkout_discount(
             ]
         )
 
+    _delete_gift_line(lines_info)
+
+
+def _delete_gift_line(lines_info: Iterable["CheckoutLineInfo"]):
+    gift_line_infos = [line for line in lines_info if line.line.is_gift]
+    if gift_line_infos:
+        gift_line_info = gift_line_infos[0]
+        gift_line_info.line.delete()
+        if gift_line_info:
+            lines_info.remove(gift_line_info)  # type: ignore[attr-defined]
+
+
+def _handle_gift_reward(
+    checkout: "Checkout",
+    checkout_info: "CheckoutInfo",
+    lines_info: Iterable["CheckoutLineInfo"],
+    gift_listing: ProductVariantChannelListing,
+    discount_object_defaults: dict,
+    rule_info: VariantPromotionRuleInfo,
+    save: bool,
+):
+    with transaction.atomic():
+        line, line_created = create_gift_line(checkout, gift_listing.variant_id)
+        line_discount = None
+        discount_created = False
+        if not line_created:
+            checkout_lines = CheckoutLine.objects.filter(checkout_id=checkout.pk)
+            line_discount = CheckoutLineDiscount.objects.filter(
```

Comment:
Can't you combine the first filter with `get_or_create`? It looks like you could just use `get_or_create` here.

Reply:
I can't as the `line` might be different and we should ensure that there is one `ORDER_PROMOTION` per checkout